std/io/bufio.ts

```diff
@@ -405,6 +406,375 @@ export class BufReader implements Reader {
+   * words, EOF is always treated as a delimiter. Once the buffer is empty,
+   * it returns `null`.
+   *
+   * Because the data returned from `readSlice()` will be overwritten by the
+   * next I/O operation, most clients should use `readString()` instead.
+   */
+  readSlice(delim: number): Uint8Array | null {
+    let s = 0; // search start index
+    let slice: Uint8Array | undefined;
+
+    while (true) {
+      // Search buffer.
+      let i = this.buf.subarray(this.r + s, this.w).indexOf(delim);
+      if (i >= 0) {
+        i += s;
+        slice = this.buf.subarray(this.r, this.r + i + 1);
+        this.r += i + 1;
+        break;
+      }
+
+      // EOF?
+      if (this.eof) {
+        if (this.r === this.w) {
+          return null;
+        }
+        slice = this.buf.subarray(this.r, this.w);
+        this.r = this.w;
+        break;
+      }
+
+      // Buffer full?
+      if (this.buffered() >= this.buf.byteLength) {
+        this.r = this.w;
+        // #4521 The internal buffer should not be reused across reads because it causes corruption of data.
+        const oldbuf = this.buf;
+        const newbuf = this.buf.slice(0);
+        this.buf = newbuf;
+        throw new BufferFullError(oldbuf);
+      }
+
+      s = this.w - this.r; // do not rescan area we scanned before
+
+      // Buffer is not full.
+      try {
+        this._fill();
+      } catch (err) {
+        err.partial = slice;
+        throw err;
+      }
+    }
+
+    // Handle last byte, if any.
+    // const i = slice.byteLength - 1;
+    // if (i >= 0) {
+    //   this.lastByte = slice[i];
+    //   this.lastCharSize = -1
+    // }
+
+    return slice;
+  }
+
+  /** `peek()` returns the next `n` bytes without advancing the reader. The
+   * bytes stop being valid at the next read call.
+   *
+   * When the end of the underlying stream is reached, but there are unread
+   * bytes left in the buffer, those bytes are returned. If there are no bytes
+   * left in the buffer, it returns `null`.
+   *
+   * If an error is encountered before `n` bytes are available, `peek()` throws
+   * an error with the `partial` property set to a slice of the buffer that
+   * contains the bytes that were available before the error occurred.
+   */
+  peek(n: number): Uint8Array | null {
+    if (n < 0) {
+      throw Error("negative count");
+    }
+
+    let avail = this.w - this.r;
+    while (avail < n && avail < this.buf.byteLength && !this.eof) {
+      try {
+        this._fill();
+      } catch (err) {
+        err.partial = this.buf.subarray(this.r, this.w);
+        throw err;
+      }
+      avail = this.w - this.r;
+    }
+
+    if (avail === 0 && this.eof) {
+      return null;
+    } else if (avail < n && this.eof) {
+      return this.buf.subarray(this.r, this.r + avail);
+    } else if (avail < n) {
+      throw new BufferFullError(this.buf.subarray(this.r, this.w));
+    }
+
+    return this.buf.subarray(this.r, this.r + n);
+  }
+}
+
```

Comment:
I'm not so happy with having a bunch of very technical duplicated code like this... it's very difficult to keep in-sync with the async version and thus quite prone to bugs.

Reply:
I wasn't super happy with that either. There are duplicated functions whose difference is one awaited call...  So two thoughts:

(1) Factoring out some common sync code is an alternative.  But there would be more function calls, some in loops.  It could possibly impact performance with function call overhead, but maybe doesn't matter too much because it's compared to IO?  If that's ok, I can do that.

(2) Keep sync and async in separate files.  I initially had them in separate files for easy diff'ing.  I saw `BufWriterSync` was previously added into `bufio.ts`, so I went the single file route.  I'd like to make them separate files if that's ok.

Reply:
It's not immediately obvious to me what the right thing to do here is. I'll try to dig into this later...

Reply:
I figured out a way to completely deduplicate all async/sync code between `BufReader`/`Sync` through an `AbstractBufReaderBase` that switches between sync/async calls... by working with promises instead of await.  It works!  How does that look?